docs/sdg/sdg-flow-yaml.md

```diff
@@ -0,0 +1,157 @@
+
+In terms of the API constructs, a Pipeline is created from a sequence of “block configurations” which express how to instantiate and invoke the individual steps (aka blocks) in the pipeline. A Flow construct serves as a template from which a sequence of block configs can be generated.
+
+## Objective
+
+- Library users can specify a custom flow using a well-defined file format.
+- Library users can either use a custom flow standalone, or combine a custom flow with existing flows.
+- The file format and library can evolve substantially without making breaking changes.
+- Incompatible changes can be introduced while retaining support for existing custom flows for a deprecation period.
+
+## Proposal
+
+We will add an API that instantiates a `Flow` object from a file:
+
+```python
+flow = Flow.from_file('mycustomflow.yaml', params)
+block_configs = flow.render()
+```
+
+The YAML file format will mirror the API and look like this:
+
+```yaml
+version: 1.0
+block_configs:
+  - block_type: LLMBlock
+    block_config: # LLMBlock constructor kwargs
+      block_name: gen_knowledge
+      output_cols: ["question", "response"]
+    gen_kwargs: # kwargs for block.generate()
+      max_tokens": 2048,
+    drop_duplicates: ["question"]
+  - block_type: FilterByValueBlock
+    block_config:
+      block_name: filter_faithfulness
+      filter_column": judgment
+      filter_value: YES
+      operation: operator.eq
+      batch_kwargs:
+        num_procs: 8
+     drop_columns: ["judgment", "explanation"]
+```
+
+## Versioning
+
+A mandatory `version` field in the YAML file expresses major and minor versions (e.g., 1.0, 1.1, 2.0).
+
+Compatibility rules
+
+1. If the major version of the YAML file is higher than the parser can handle, the parser should reject the file.
+2. If the minor version of the YAML file is higher than the highest version the parser is aware of, the parser should read the file but ignore any unrecognized content.
+3. If the file’s version is lower than the parser version, the parser should provide default values for any configuration introduced in later versions.
+
+Example parsing logic:
+
+```python
+def parse_custom_flow(content):
+    version = content['version']
+    major, minor = map(int, version.split('.'))
+
+    if major > PARSER_MAJOR:
+        raise IncompatibleVersionError("The custom flow file format is from a future major version.")
+    elif major <= PARSER_MAJOR and minor > PARSER_MINOR:
+        logger.warning("The custom flow file may have new features that will be ignored.")
+```
+
+### Flow Params
+
+The following parameters are supplied to a Flow constructor and used by the render() method:
+
+- client - an OpenAI completions API client for talking to the teacher model via the serving backend (i.e.  llama-cpp or vLLM)
+- model_family - e.g. mixtral or merlinite
+- model_id - a path name for the specific teacher model being used
+- num_iters - how many samples to generate
```

Comment:
I'm not sure if `num_iters` conveys well its description `how many samples to generate`.

Reply:
Definitely in the realm of API change here, but #77 will change it to `num_instructions_to_generate` :+1: